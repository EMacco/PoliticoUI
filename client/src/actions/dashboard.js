import axios from 'axios';
import { toast } from 'react-toastify';
import {
  UPDATE_PROFILE_PICTURE,
  SET_INTERESTED_OFFICE,
  GET_ERRORS,
  SET_USER_VOTES
} from '@actions/types';
import { fetchOfficeDetailsByID, collateResult } from './office';
import { fetchUserByID } from './user';
import { fetchPartyByID } from './party';

const fetchInterestedOffice = userDetails => async dispatch => {
  try {
    const { data } = (await axios.get('/offices/candidates')).data;
    for (let ind = 0; ind < data.length; ind += 1) {
      if (data[ind].candidateid === userDetails.user.id) {
        const fullData = (await fetchOfficeDetailsByID(data[ind].officeid)).data.data[0];
        let numberOfCandidates = 0;
        for (let ind2 = 0; ind2 < data.length; ind2 += 1) {
          if (data[ind2].officeid === data[ind].officeid) {
            numberOfCandidates += 1;
          }
        }

        let electionDate = data[ind].date;
        if (!electionDate) {
          electionDate = 'Not set';
        } else {
          electionDate = new Date(electionDate).toDateString();
        }
        const payload = {
          image: fullData.logourl,
          officeId: fullData.id,
          officeName: fullData.name,
          numberOfCandidates,
          electionDate
        };

        dispatch({ type: SET_INTERESTED_OFFICE, payload });
        break;
      }
    }
  } catch (err) {
    dispatch({ type: GET_ERRORS, payload: { global: err.response.data.error } });
  }
};

const fetchPageVotes = ({ id }) => async dispatch => {
  try {
    const { dataa } = (await axios.get(`/office/${id}/user-votes`)).data;
    let votes = [];
    for (let ind = 0; ind < dataa.length; ind += 1) {
      const userRes = await fetchUserByID(dataa[ind].candidateid);
      const userDets = userRes.data;
      const name = `${userDets.data[0].firstname} ${userDets.data[0].lastname}`;
      const officeId = dataa[ind].officeid;
      const partyId = userDets.data[0].partyid;
      const myCandidateId = dataa[ind].candidateid;
      let officeName = '';
      let partyName = '';
      let partyLogo = '';
      let result = 0;
      let resultStyle = 'won';
      let resultString = 'won';
      const officeResp = await fetchOfficeDetailsByID(officeId);
      const officeRes = officeResp.data;
      officeName = officeRes.data[0].name;
      
      const partyResp = await fetchPartyByID(partyId);
      const partyRes = partyResp.data;
      partyName = partyRes.data[0].name;
      partyLogo = partyRes.data[0].logourl;
      
      const resultResp = await collateResult(officeId);
      const resultRes = resultResp.data;
      let winningCandidate;
      let winningCandidateVote = 0;
      for (let rInd = 0; rInd < resultRes.data.length; rInd += 1) {
        if (resultRes.data[rInd].candidateid === myCandidateId) {
          result = resultRes.data[rInd].count;
        }

        if (resultRes.data[rInd].count > winningCandidateVote) {
          winningCandidate = resultRes.data[rInd].candidateid;
          winningCandidateVote = resultRes.data[rInd].count;
        }

        if (winningCandidate === myCandidateId) {
          resultStyle = 'won';
          resultString = 'won';
        } else {
          resultStyle = 'lost';
          resultString = 'lost';
        }

        const today = new Date().toDateString();
        const elecDate = new Date(dataa[ind].createdon).toDateString();
        if (today === elecDate) {
          if (resultString === 'won') {
            resultString = 'wining';
          } else if (resultString === 'lost') {
            resultString = 'losing';
          }
        }
      }

      const votesObject = {
        partyLogo,
        name,
        partyName,
        officeName,
        result,
        resultStyle,
        resultString
      };
      votes.push(votesObject);
      if (ind === dataa.length - 1) {
        dispatch({ type: SET_USER_VOTES, payload: votes });
      }
    }
  } catch (err) {
    dispatch({ type: GET_ERRORS, payload: { global: err.response.data.error } });
  }
};

export { fetchInterestedOffice, fetchPageVotes };
