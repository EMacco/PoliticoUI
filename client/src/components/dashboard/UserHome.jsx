import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import AuthHeader from '@components/layouts/AuthHeader';
import AuthFooter from '@components/layouts/AuthFooter';
import AuthNav from '@components/layouts/AuthNav';
import Card from '../layouts/Card';
import { fetchInterestedOffice, fetchPageVotes } from '@actions/dashboard';

class UserHome extends Component {
  componentDidMount() {
    const { user } = this.props.auth;
    this.props.fetchInterestedOffice({ user });
    this.props.fetchPageVotes(user);
  }

  render() {
    const { votes, interest } = this.props;
    const { officeId, officeName, image, numberOfCandidates, electionDate } = interest;

    const interestedOfficeContent = [
      {
        officeId,
        officeName,
        numberOfCandidates,
        type: 'link'
      },
      {
        title: 'Date',
        value: electionDate
      }
    ];

    let interestCard = '';

    if (officeId !== undefined) {
      interestCard = (
        <Fragment>
          <Card name={officeName} image={image} content={interestedOfficeContent} />
          <Card hidden="true" />
        </Fragment>
      );
    }

    let votesCards = [];
    this.props.votes.map((vote, ind) => {
      const { partyLogo, name, partyName, officeName, result, resultStyle, resultString } = vote;

      const voteContent = [
        {
          title: 'Party',
          value: partyName
        },
        {
          title: 'Office',
          value: officeName
        },
        {
          title: 'Result',
          value: result
        },
        {
          text: resultString,
          style: `${resultStyle}-election`,
          type: 'span'
        }
      ];

      votesCards.push(
        <Fragment key={Math.random().toString()}>
          <Card name={name} image={partyLogo} content={voteContent} />
          {votes.length % 2 !== 0 && ind === votes.length - 1 ? <Card hidden="true" /> : ''}
        </Fragment>
      );
    });

    return (
      <div>
        <AuthHeader />
        <main id="user-home-main">
          <div className="container">
            <AuthNav />

            <section id="user-home-content-div">
              <div id="loadingIconContainer" />
              <div className="profile-page-section">
                <h3 id="interested-office-header">Interested Office</h3>
                <div className="content-group" id="interestedOfficeSlot">
                  {interestCard}
                </div>
              </div>

              <div className="profile-page-section">
                <h3>My Votes ({votes.length})</h3>
                <div className="content-group" id="userVotesSlot">
                  {votesCards}
                </div>
              </div>
            </section>
          </div>
          <AuthFooter />
        </main>
      </div>
    );
  }
}

UserHome.propTypes = {
  fetchInterestedOffice: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
  interest: PropTypes.object.isRequired,
  fetchPageVotes: PropTypes.func.isRequired,
  votes: PropTypes.array.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth,
  interest: state.dashboard.interests,
  votes: state.dashboard.votes
});

export default connect(
  mapStateToProps,
  { fetchInterestedOffice, fetchPageVotes }
)(UserHome);
