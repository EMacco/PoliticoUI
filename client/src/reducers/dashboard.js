import { SET_INTERESTED_OFFICE, SET_USER_VOTES, CLEAR_DASHBOARD, GET_ERRORS } from '../actions/types';

const initialState = {
  interests: {},
  votes: []
};

const dashboardReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_INTERESTED_OFFICE:
      return { ...state, interests: action.payload };
    case SET_USER_VOTES:
      return { ...state, votes: action.payload };
    case CLEAR_DASHBOARD:
        return { ...state, interests: {}, votes: [] };
    default:
      return state;
  }
};

export default dashboardReducer;
