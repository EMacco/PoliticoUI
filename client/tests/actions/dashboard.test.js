import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { fetchInterestedOffice, fetchPageVotes } from '@actions/dashboard';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const store = mockStore({
  auth: {},
  dashboard: {},
  parties: {},
  offices: {}
});

const userDetails = {
  user: {
    id: 1,
    firstname: 'John',
    lastname: 'Doe',
    email: 'testing@testing.com',
    phone: '08088293829',
    password: 'password'
  }
};

describe('Dashboard actions', () => {
  it('should dispatch SET_INTERESTED_OFFICE on successful fetch', async () => {
    await store.dispatch(fetchInterestedOffice(userDetails));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should dispatch SET_USER_VOTES on successful fetch of user votes', async () => {
    await store.dispatch(fetchPageVotes({ id: 1 }));
    expect(store.getActions()).toMatchSnapshot();
  });
});
