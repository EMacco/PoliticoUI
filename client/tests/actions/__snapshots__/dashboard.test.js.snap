// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard actions should dispatch SET_INTERESTED_OFFICE on successful fetch 1`] = `
Array [
  Object {
    "payload": Object {
      "electionDate": "Fri Mar 01 2019",
      "image": "https://res.cloudinary.com/dn4pokov0/image/upload/v1551452411/mtuaqvskpzlwlu331g1e.png",
      "numberOfCandidates": 1,
      "officeId": 1,
      "officeName": "office of the president",
    },
    "type": "SET_INTERESTED_OFFICE",
  },
]
`;

exports[`Dashboard actions should dispatch SET_USER_VOTES on successful fetch of user votes 1`] = `
Array [
  Object {
    "payload": Object {
      "electionDate": "Fri Mar 01 2019",
      "image": "https://res.cloudinary.com/dn4pokov0/image/upload/v1551452411/mtuaqvskpzlwlu331g1e.png",
      "numberOfCandidates": 1,
      "officeId": 1,
      "officeName": "office of the president",
    },
    "type": "SET_INTERESTED_OFFICE",
  },
  Object {
    "payload": Array [
      Object {
        "name": "John Doe",
        "officeName": "office of the president",
        "partyLogo": "https://res.cloudinary.com/dn4pokov0/image/upload/v1551452220/ukhkvobjnabok4ymt7g3.jpg",
        "partyName": "all progressives congress",
        "result": 0,
        "resultString": "lost",
        "resultStyle": "lost",
      },
    ],
    "type": "SET_USER_VOTES",
  },
]
`;
